File: ruby/find_first_duplicate.rb
def find_first_duplicate(arr)
  if arr.empty? || arr.length == 1
    return -1
  end

  temp = []
  i = 0
  while !temp.include?(arr[i]) do
    temp.push(arr[i])
    i+=1
  end

  i < arr.length ? arr[i] : -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  puts

  puts "Expecting 4"
  puts "=>", find_first_duplicate([2,1,5,4,7,4,5,1])
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  if (!arr.length || arr.length === 1) {
    return -1
  }

  const temp = []
  let indexFocus = 0
  while (!temp.includes(arr[indexFocus])) {
    temp.push(arr[indexFocus])
    indexFocus ++
  }
  return indexFocus < arr.length ? arr[indexFocus] : -1
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file

// create temporary array
// push numbers from original array into temp array
// until temp array contains current number
// if current number index < original array length
// return last number of temp array
// else 
// return -1

// And a written explanation of your solution

// This function will cycle through the original array until a duplicate number is found.
// If no duplicate is found, the function will return -1, otherwise it will return the number of current focus.